### ISSUE TITLE
feature request: webhooks support

### ISSUE THREAD

**Issue by capcom6** [m-adawi/swarm-cd#47]
## Overview
Pulling changes by timer is not always the most efficient approach. Some services may update very infrequently, perhaps once a week or even less frequently.

To address this, it would be beneficial to add support for webhooks. Triggering the webhook endpoint would initiate an update of the stack.

## Implementation Details
Webhook configuration should be implemented at the repository level. This feature could also serve as a foundation for manually triggering stack updates from the Web UI.

## Related Issues
- [Web UI](https://github.com/m-adawi/swarm-cd/issues/16)

**Comment by EhabY** [comment]
The timer approach is indeed inefficient but it was a way to get the UI up and running quickly. Triggering an update when an event from the server is received makes more sense. Though we probably have to batch these events in the server, in case they happen too quickly.

There are two ways straightforward ways to do this:

1. Websockets (duplex communication) - Which would make using other endpoints redundant since all communication can happen in the socket.
2. Server-Send Events (SSE) - Unidirectional communication.

SSEs make more sense to me since it's just a notification sent from the server to the client, and we can reuse already existing endpoints that other clients use (like a CLI) to request the new content.

**Comment by capcom6** [comment]
Sorry, I wasn't clear. I mean using webhooks to react to changes in a repository containing a `docker-compose.yml` file.

**Comment by sanzoghenzo** [comment]
This would be a nice feature to have!

Which events do you think we should handle? Is the push event enough?

Some remarks:

- each git service has its own webhooks with different payloads and headers, see [github](https://docs.github.com/en/webhooks/webhook-events-and-payloads) and [gitlab](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#push-events), they may be similar but we have to take into account that the info we need can be in different places
- adding support for this, and for an API in general, would mean a complete rewrite of the core, since it is a simple loop over the configured repositories to check for changes.

@capcom6 following the discussion in #46: I don't want to discourage you in any way, and this is still my personal opinion, but it seems to me that this project is not what you need after all 😅 
Let's see if @m-adawi proves me wrong!

**Comment by sanzoghenzo** [comment]
Thanks for the explanation; I never used webhooks so I thought it would need a single API endpoint + headers and payload parsing to determine the stacks to update 😅  it is definitely easier to implement it like that

**Comment by m-adawi** [comment]
I think a webhook is useful indeed. It can be integrated with GitHub Apps to make github send a request to SwarmCD informing it of new stack updates instead of having SwarmCD periodically pull for changes. But this also has security concerns, do you really want to publicly expose a tool that has full control of your docker socket? 

Private webhooks can be useful too, for example if I want to press a button in the UI to make SwarmCD update a stack now instead of waiting the pull interval

**Comment by oliverlambson** [comment]
+1 for webhooks, think they would be a huge boost (the whole reason I came across swarmCD is because I’m looking for “ArgoCD for swarm”)

**Comment by m-adawi** [comment]
@oliverlambson Yes PRs are welcome! Feel free to submit any contributions. Looking forward to it!

**Comment by Spliterash** [comment]
+1, that would be nice to use github webhook's
